16.0 mig urban mine

Honeyxilia opened the pull request.

fcayre committed:
Import the urban_mine module into the 12.0 branch

legalsylvain committed:
[ADD] migration analysis files

legalsylvain committed:
[REF] remove some modules from the CI, to avoid to install dependencies.
[REF] only ignore 2 modules, with failing tests.
[FIX] commown_cooperative_campaign: correct version.
[FIX] magic/python-magic installation. Use : external_dependencies_override
[FIX] test-requirements.txt : fix typo, and install 'account_payment_slimpay'
[REF]  urban_mine : set installable to False

legalsylvain committed:
[REF] apply pre-commit, fixing website values

fcayre committed:
Import v10 version

fcayre committed:
First installable version

Not that report.xml is not read.

fcayre committed:
Refactor urban mine registration actions into a res.partner method

fcayre committed:
Refactor urban mine send label action into a crm.lead method

fcayre committed:
Refactor urban mine payment related actions into a lead method

... and make tests all pass.

fcayre committed:
Add the api.returns decorator to the create api methods

legalsylvain committed:
[META] enable black

legalsylvain committed:
[META] enable isort

legalsylvain committed:
[META] enable black (new version)

fcayre committed:
Change .fr to .coop in mail templates

Those templates are in noupdate mode so this should not hurt.

github-grap-bot committed:
urban_mine 12.0.1.0.1

fcayre committed:
Fix label class in website forms

col-form-label replaced control-label.

fcayre committed:
Adapt auto-invoice py3o report to v12

github-grap-bot committed:
urban_mine 12.0.1.0.2

fcayre committed:
urban_mine 12.0.1.0.3

fcayre committed:
Base urban mine upon project instead of CRM and use 2 different campaigns

First campaign is the same as previous one: buy the phone and give a
coupon. Second campaign only sends a coupon, with a higher reduction.

The customer has to choose the campaign, so there is an extract stage
to ask if necessary.

Also add phone as required in the registration web page, as Colissimo
requires it. The registration page is now updatable as it should not
be modified too much in the UI.

fcayre committed:
Replace recaptcha by hcaptcha in urban_mine registration page (and all similar forms)

fcayre committed:
urban_mine 1.1.0

fcayre committed:
Remove automatic payment of urban mine auto-invoices

They will be reconciled when the corresponding bank statement will be imported.

fcayre committed:
Update urban mine product price

fcayre committed:
Use adapted payment term for urban mine auto-invoices

fcayre committed:
Improve label attachment name in urban mine label email

fcayre committed:
Do not pass urban mine product to a method that can use its ref to get it

fcayre committed:
Add a bit of documentation to urban mine's readme

fcayre committed:
Factor out small code to name a urban mine task

fcayre committed:
Automatically generate a PO when a urban mine device is validated

The purchased product must be specified in the project task beforehand.

fcayre committed:
Fix wrong product id used for the invoice

fcayre committed:
Add a step in urban mine workflow and document it

fcayre committed:
i18n update

fcayre committed:
urban_mine 1.1.1

fcayre committed:
Show country and add urban-mine currently supported options

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
[security] Do not perform important actions as someone else

Sudo was used abusively here and could be used to perform important actions
(like payments) in the name of anyone as soon as you could assign that person
to the task, which was highly dangerous and not justified in any ways...

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Add a parcel type specific to urban mine

... to avoid deletion.

fcayre committed:
[manual] post-merge updates

Merged manually because the bot crashes when generating the readme and we will
soon migrate so it is useless to write a readme now with the old 12.0 format.

fcayre committed:
Add missing i18n pot files

Honeyxilia committed:
[IMP] urban_mine: black, isort, prettier

Honeyxilia committed:
[MIG] urban_mine: Migration to 16.0

Honeyxilia committed:
Removed depreciated @api.returns decorator

Honeyxilia committed:
Changed dependency for modules merged into website

website_form_builder and website_form_recaptcha have been merged into website_form, and then into website, so we now depend on that module.

Honeyxilia committed:
[test] Replaced at/post_install decorators by tagged

Honeyxilia committed:
Migrated Jinja mail templates to QWeb

Honeyxilia committed:
Adapted values to account.invoice -> account.move migration

Honeyxilia committed:
Adapted res.partner creation to allow batch creation

Honeyxilia committed:
[test] Replaced setUp method by setUpClass

Honeyxilia committed:
[test] Changed supplier declaration for partner creation

In the account model, the supplier boolean value has been replaced by supplier_rank, an Integer ranking partners by their amount of sales.
A partner with a supplier_rank of 1 or more is considered a supplier, so we set the value to 1.

Honeyxilia committed:
[test] Removed the depreciated 'name' alias by args[0]

Honeyxilia committed:
Added sequence code to stock.picking.type record

Honeyxilia committed:
Changed expense account retrieval

Retrieving the expense account from its reference is trickier in Odoo 16, because the accounts are created from template, with the reference based on the id of the company (<company id.>_pcg_6068).
So, we retrieve the expense account by fetching the code of the expense account - 606800.

Honeyxilia committed:
Changed migrated user_id field to user_ids

Honeyxilia committed:
Adapted to ir.actions.report render method change

Honeyxilia committed:
[test] Added account reloading for expanse account

In a local installation, the environnement company (base.main_company) doesn't match the company for which the l10n_fr records are created. This is mainly problematic for the expense account of the Rental Fairphone product template, which crashes the account.move post action.
So, to pass tests in a local installation, we reload the l10n_fr records for the base company.

Honeyxilia committed:
Added invoice_date field set in account.move creation

To be able to perform the post action on a purchase invoice, it's required that the invoice_date is set. Since the related purchase.order's purchase date is also set to the day of the execution of the method, we also set the invoice date to the current date on which the invoice is created.

Honeyxilia committed:
Reimplemented hCaptcha widget through website_recaptcha_v2 module

- Since the base implementation of reCAPTCHA in the Odoo module is based on reCAPTCHA v3, and the v3 implementation of hCaptcha is kept behind an Enterprise subscription, we aimed to reimplement hCaptcha (reCAPT. v2) through other means.
- While researching, we found the OCA module 'website_recapatcha_v2', which reimplements reCAPTCHA v2 on Odoo 16 - so, urban_mine now depends on website_captcha_v2 to reimplement hCaptcha.
(Since the values for reCAPTCHA were hardcoded in the module, we added the ability to customized the CAPTCHA API used in website_recaptche_v2 : https://github.com/OCA/website/pull/1111 )
- We then call the recaptcha_widget template to insert and initialize the hCaptcha check, by changing the values in Odoo's settings.

Honeyxilia committed:
Adapted odt report to account.invoice -> move migration

Details :
- Deposit Account retrieval by code rather than external id
- type -> move_type
- ('open', 'paid') -> 'posted'
- refund_invoice_id -> reversal_move_id
- number/move_name -> name
- incoterms_id -> invoice_incoterm_id
- uom_id -> product_uom_id
- Show categories of taxes with tax_totals
- payment_term_id -> invoice_payment_term_id
- added html_sanitization around invoice_payment_term_id.note

NOTES:
I translated 'Total Taxes' by 'Total TVA' in the recap - it might need to be changed depending on the translation of Total Taxes in custom_report.

Honeyxilia committed:
Rearranged report-related files in structure

Honeyxilia committed:
Remade Urban Mine form with Odoo 16 page editor

Since the website_form_builder module has been fused in the website module, the format of built forms has been changed, so some values weren't correctly parsed when creating a partner, and the form itself wasn't recognized by the editor, making its modification difficult.
So, we reimplemented the form with the new editor.

We also took the opportunity to change the 'supplier' field to 'supplier_rank', with a base rank of 1.

Honeyxilia committed:
Migrated Incoterms field in custom_report report

While migrating the urban_mine report, I realized that I had missed the Incoterms field, which has had its name changed (incoterms_id -> invoice_incoterm_id).
So, I modified it here.

Honeyxilia committed:
Changed black/whitelisting of res_partner fields for website forms

- We changed whitelisted fields 'surname' and 'supplier', to 'lastname' and 'supplier_rank' respectively.
- We removed the depreciated field attribute website_form_blacklisted

Honeyxilia committed:
Added reCAPTCHA v2 validation for website form

At the moment, website_recaptcha_v2 doesn't implement any form validation controllers for their widget. However, it still adds a website method to check with the verification endpoint of the CAPTCHA API.
So, we use that method to validate or invalidate the CAPTCHA in the form.

NOTE: in the website_form overload, we do precise the route type - this is because in the test, the type isn't considered as set, so we add the argument overload - however, the original method still has its type set to 'http'.

Honeyxilia committed:
[test] Added demo data for hCaptcha testing

To be able to test the hCaptcha widget on a local Odoo instance, we add test keys and hCaptcha related values to be able to locally load hCaptcha, and then be able to test the form validation.
We also add base hCaptcha configuration values for convinience

Honeyxilia committed:
[test] Improve cov.: Changed assigned user of task

In the urban_mine_send_mail method, if the current env. user is not in the task assignees, then the method should perform the action as one of the task assignees.

Honeyxilia committed:
[test] Added OCA/website PR to test sources

With our PR to add captcha API customization to OCA/website (https://github.com/OCA/website/pull/1111), we added fields in the website model, which we use in the urban_mine module.
So, we add this source in our test requirements in order to fetch the code in the GitHub test job.

fcayre commented on urban_mine/models/project_task.py:
```diff
@@ -76,28 +75,23 @@ def urban_mine_build_autoinvoice(
                             "product_id": product.id,
                             "name": description,
-                            "account_id": product.property_account_expense_id.id,
-                            "analytic_tag_ids": [(6, 0, tags.ids)] if tags else False,
```
analytic_distribution remplace analytic_tag_ids je pense. Un exemple dans product_rental/models/sale_order.py (qu'il faut transposer à account.move.line)

Honeyxilia replied:
En effet - j'ai donc créé un analytic account avec le·a même partenaire que la carte, et le plan de la compagnie de l'environnement c:
```py
        aa = self.env["account.analytic.account"].create(
            {
                "name": self.urban_mine_name(),
                "partner_id": self.partner_id.id,
                "plan_id": self.env.company.analytic_plan_id.id,
            }
        )
```

Honeyxilia replied:
Retranscription de nos discussions sur Odoo: Nous avons finalement décidé de réinstaurer les étiquettes analytiques (account.analytic.tag) via le nouveau module OCA, account_analytic_tag

fcayre commented on urban_mine/data/product.xml:
```diff
@@ -10,7 +10,7 @@
       <field name="standard_price">60</field>
       <field name="invoice_policy">order</field>
-      <field name="property_account_expense_id" ref="l10n_fr.1_pcg_6068" />
+      <field name="property_account_expense_id" search="[('code','=', '606800')]" />
```
Typo message de commit : tricker => trickier

fcayre replied:
Intéressant, je ne connaissais pas cette syntaxe `search=`, bien vu !
En revanche je suis embêté si on avait (pas le cas ici) plusieurs companies françaises (donc avec chacune un compte avec le code 606800), ça ne fonctionnerait pas...

Honeyxilia replied:
En effet - dans ce cas là, il faudrait potentiellement baser le compte sur le nom/identifiant de la compagnie, mais il faudrait gérer les cas pour les environnements où FR Company ou Commown serait la compagnie dans l'environnement...

fcayre commented on urban_mine/data/mail_templates.xml:
```diff
@@ -7,16 +7,16 @@
-        >${object.company_id.name} - Accusé Réception offre de reprise de votre Fairphone - Référence COMMOWN-MU-${object.id}</field>
-      <field name="partner_to">${object.partner_id.id}</field>
+        >{{object.company_id.name}} - Accusé Réception offre de reprise de votre Fairphone - Référence COMMOWN-MU-{{object.id}}</field>
+      <field name="partner_to">{{object.partner_id.id}}</field>
```
Est-ce que tu utilises un script basé sur ta PR sur openupgrade pour faire ces modifs de manière automatisée (et donc plus sûre) ? Si oui je veux bien une recette, perso je fais à la main et comme tu le sais il m'est arrivé de me tromper.
Si non pas grave, on finit comme ça !

Honeyxilia replied:
Non, je l'ai fait à la main... heureusement, il n'y a pas tant de changements que ça à appliquer, mais c'est vrai que ça aurait été plus efficace avec un script

fcayre commented on urban_mine/report/report_autoinvoice.odt:
Le compte récupéré par code, il faut idéalement ajouter un critère sur company_id aussi (genre `("company_id", "=", env.company.id)` ou approchant). Mais je ne suis pas encore au clair là-dessus, ça mériterait qu'on fasse des tests pour bien comprendre comment ça fonctionne les companies, car pour l'instant on n'est pas concerné, mais ça pourrait venir.

fcayre replied:
Pour ce qui concerne "open", "paid", tu es sûre qu'il ne faut pas plutôt utiliser payment_state et les valeurs "paid", "not_paid" ?

Honeyxilia replied:
Concernant posted, la valeur est basée sur la valeur state, qui permet de dire - pour le titre de la facture -  si la facture est un brouillon (`draft`), annulée  (`cancel`) ou postée (`posted`).
Ainsi, je ne pense pas qu'il soit pertinent de changer la valeur pour déterminer le titre de la facture

Honeyxilia replied:
Concernant le compte de dépôt, on peut rajouter une clause `property_account_income_categ_id.company_id == o.env.company` c:

fcayre replied:
OK pour les 2.

fcayre commented on urban_mine/data/registration_page.xml:
```diff
@@ -14,378 +14,376 @@
               <div class="container">
                 <div class="row">
                   <div class="col-lg-12">
-                    <h1 class="text-center">Revendez votre appareil à la coopérative<br
```
Typo dans le message de commit : recoignized => recognized et suplier => supplier

Honeyxilia replied:
Fait c:

fcayre commented on urban_mine/data/registration_page.xml:
```diff
@@ -14,378 +14,376 @@
                     <h3
                                             class="text-center text-muted"
-                                        >en remplissant le formulaire suivant&amp;nbsp;:</h3>
+                                        >en remplissant le formulaire suivant:</h3>
```
En typographie française on met une espace insécable de ":", ce qui correspond à l'entité html &nbsp; qui donne le code ci-dessus avec l'échappement xml. Pour simplifier tu peux cependant copier-coller une espace insécable, en la cherchant sur wikipedia par exemple.

Honeyxilia replied:
Fait

fcayre commented on urban_mine/__manifest__.py:
```diff
@@ -15,10 +15,10 @@
         "report_py3o",
         "website",
         "website_sale_coupon",
+        "website_recaptcha_v2",
```
Typo dans le message de commit : though other means => through other means

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,62 @@
+
+        # A partner should be created and its id returned with the response.
+        self.assertTrue(resp_ok_data.get("id"))
+        self.assertTrue(self.env["res.partner"].browse(resp_ok_data["id"]))
```
`.browse()` n'échoue pas quand l'id n'est pas trouvé en base, car il ne fait pas de requête pour vérifier, par souci de performance. Il faut donc rajouter `.exists()` pour que ton test soit pertinent.

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,62 @@
+        self.partner_data.update(
+            {
+                "h-captcha-response": self.test_hcaptcha_success_key,
+            }
```
Pffiuuu, j'aurais condensé ça un peu sans perte de lisibilité : 
```
        self.partner_data.update(
            {"h-captcha-response": self.test_hcaptcha_success_key}
        )
```

voire en renommant `test_hcaptcha_success_key` en `hcaptcha_success_key` : 
```
        self.partner_data.update({"h-captcha-response": self.hcaptcha_success_key})
```

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,62 @@
+            req.params = self.partner_data
+            resp_no_token = UrbanMineWebsiteForm().website_form("res.partner")
+
+        resp_no_token_data = json.loads(resp_no_token.data.decode("utf-8"))
```
le `decode("utf-8")` est vraiment nécessaire ? J'ai testé avec succès `json.loads(b'["1", "2"]')` comme `json.loads('["1", "2"]')` donc je serais étonné.

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,62 @@
+        self.assertFalse(resp_ok_data.get("error"))
+
+        # A partner should be created and its id returned with the response.
+        self.assertTrue(resp_ok_data.get("id"))
```
J'aurais utilisé le plus expressif : `self.assertNotIn(...)` (idem plus bas)

fcayre replied:
`self.assertTrue(resp_ok_data.get("id"))` s'écrit plus joliment `self.assertIn("id", resp_ok_data)`

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,62 @@
+        self.assertIn(
+            "Captcha failed",
+            resp_no_token_data["error"],
+        )
```
À condenser en une ligne ?

Honeyxilia committed:
Added OCA module account_analytic_tag dependency

Since Odoo 16.0, the account.analytic.tag model and its related field in account.move.line have been taken out of the official Odoo modules, and they were reimplemented as an OCA module ( https://github.com/OCA/account-analytic/tree/16.0 ).
As such, we now depend on that module to apply analytic tags on invoices (account.move)

fcayre commented on urban_mine/tests/test_controllers.py:
```diff
@@ -0,0 +1,55 @@
+        self.assertNotIn("id", resp_no_token_data)
+
+        # An error message should be returned.
+        self.assertTrue(resp_no_token_data.get("error"))
```
Idem, assertIn allège la lecture

Honeyxilia committed:
[test] Added Expl. tag test

Through action_registration_payment action, we add a tag named "EXPL" to created invoices, in order to count the invoice as part of the company's spendings with the urban mine.
Currently, the tag was not added through any module data, but directly through the web UI.
So, we create a demo tag with the same name (EXPL) to test whether the tag was correctly added to the invoice record.

fcayre commented:
/ocabot merge patch
